druid/src/widget/aspect_ratio_box.rs

```diff
@@ -0,0 +1,164 @@
+    fn generate_constraints(&self, bc: &BoxConstraints) -> BoxConstraints {
+        let (mut new_width, mut new_height) = (bc.max().width, bc.max().height);
+
+        if new_width == f64::INFINITY {
+            new_width = new_height * self.ratio;
+        } else {
+            new_height = new_width / self.ratio;
+        }
+
+        if new_width > bc.max().width {
+            new_width = bc.max().width;
+            new_height = new_width / self.ratio;
+        }
+
+        if new_height > bc.max().height {
+            new_height = bc.max().height;
+            new_width = new_height * self.ratio;
+        }
+
+        if new_width < bc.min().width {
+            new_width = bc.min().width;
+            new_height = new_width / self.ratio;
+        }
+
+        if new_height < bc.min().height {
+            new_height = bc.min().height;
+            new_width = new_height * self.ratio;
+        }
+
+        BoxConstraints::tight(bc.constrain(Size::new(new_width, new_height)))
+    }
+}
+
+/// Clamps the ratio between 0.0 and f64::MAX
+/// If ratio is 0.0 then it will return 1.0 to avoid creating NaN
+fn clamp_ratio(mut ratio: f64) -> f64 {
+    ratio = f64::clamp(ratio, 0.0, f64::MAX);
+
+    if ratio == 0.0 {
+        warn!("Provided ratio was <= 0.0.");
+        1.0
+    } else {
+        ratio
+    }
+}
+
+impl<T: Data> Widget<T> for AspectRatioBox<T> {
+    #[instrument(
+        name = "AspectRatioBox",
+        level = "trace",
+        skip(self, ctx, event, data, env)
+    )]
+    fn event(&mut self, ctx: &mut EventCtx, event: &Event, data: &mut T, env: &Env) {
+        self.inner.event(ctx, event, data, env);
+    }
+
+    #[instrument(
+        name = "AspectRatioBox",
+        level = "trace",
+        skip(self, ctx, event, data, env)
+    )]
+    fn lifecycle(&mut self, ctx: &mut LifeCycleCtx, event: &LifeCycle, data: &T, env: &Env) {
+        self.inner.lifecycle(ctx, event, data, env)
+    }
+
+    #[instrument(
+        name = "AspectRatioBox",
+        level = "trace",
+        skip(self, ctx, old_data, data, env)
+    )]
+    fn update(&mut self, ctx: &mut UpdateCtx, old_data: &T, data: &T, env: &Env) {
+        self.inner.update(ctx, old_data, data, env);
+    }
+
+    #[instrument(
+        name = "AspectRatioBox",
+        level = "trace",
+        skip(self, ctx, bc, data, env)
+    )]
+    fn layout(&mut self, ctx: &mut LayoutCtx, bc: &BoxConstraints, data: &T, env: &Env) -> Size {
+        bc.debug_check("AspectRatioBox");
+
+        if bc.max() == bc.min() {
+            warn!("Box constraints are tight. Aspect ratio box will not be able to preserve aspect ratio.");
+
+            return self.inner.layout(ctx, &bc, data, env);
+        }
+
+        if bc.max().width == f64::INFINITY && bc.max().height == f64::INFINITY {
+            warn!("Box constraints are INFINITE. Aspect ratio box won't be able to choose a size because the constraints given by the parent widget are INFINITE.");
```

Comment:
This is another case where I'm not sure the warning is always indicating "incorrect" behavior, because infinite box constraints are sometimes desirable. That said, I think this warning is far less likely to be a false positive than the one above.

Reply:
So, generally, we've been implementing warnings like this in widgets where the infinite constraint creates an impossible situation. This means Flex (sometimes, if there is a flexible child), Split, and probably others I'm forgetting. So I think it makes sense here. The warning is not that an infinite constraint is *always wrong*, but that it causes a situation the widget having to handle it cannot solve.